### PR TITLE
Enable the use of pyproject-fmt

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

Now that #15247 has been merged we may want to consider standardizing the formatting of our `pyproject.toml` file. As noted in https://github.com/astropy/astropy/pull/15247#issuecomment-1713530445 the [`pyproject-fmt`](https://github.com/tox-dev/pyproject-fmt) tool has been created by the tox devs in order to apply consistent formatting to to their `pyproject.toml` files across their various projects.

This PR is applies the `pyproject-fmt` tool to `astropy` and enables its continued use via our `pre-commit`.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
I am opening this PR so that we can discuss the pros/cons of using this.

**Comment by nstarman** [comment]
I'd like to only think about the contents of this file, not how it's formatted, so I'm in favor of this PR.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -111,6 +111,12 @@ repos:
     hooks:
       - id: black
 
+  - repo: https://github.com/tox-dev/pyproject-fmt
+    rev: "1.0.0"
+    hooks:
+      - id: pyproject-fmt
+        args: ["--indent=4"]
```
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:118]: Unfortunately, I cannot figure out how to independently configure the indent setting (requested by @pllim) outside of specifying it via command line args. I have opened tox-dev/pyproject-fmt#130 for assistance with this.

File: pyproject.toml [hunk 2]
```diff
@@ -1,110 +1,110 @@
...
     "Intended Audience :: Science/Research",
     "License :: OSI Approved :: BSD License",
     "Operating System :: OS Independent",
     "Programming Language :: C",
     "Programming Language :: Cython",
-    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
```
Comment (nstarman) [inline pyproject.toml:36]: Will this tool auto-add python3.12?

Comment (WilliamJamieson) [reply pyproject.toml:36]: It should according to https://pyproject-fmt.readthedocs.io/en/latest/#calculating-max-supported-python-version

Comment (nstarman) [reply pyproject.toml:36]: Now this is nice

**Comment by pllim** [comment]
Maybe put this in the next dev telecon agenda? I am neutral and I am interested in what other devs think.

File: pyproject.toml [hunk 3]
```diff
@@ -1,110 +1,110 @@
...
     "sphinx-astropy[confv2]>=1.9.1",
-    "pytest>=7.0",
-    "sphinx-changelog>=1.2.0",
+    "sphinx-changelog>=1.2",
     "sphinx_design",
-    "Jinja2>=3.0",
-    "tomli; python_version < '3.11'",
+    'tomli; python_version < "3.11"',
+]
+recommended = [
```
Comment (mhvk) [inline pyproject.toml:85]: Order again had (some) meaning here. That said, in this case I can see the advantage of just being alphabetical.

File: pyproject.toml [hunk 4]
```diff
@@ -407,6 +265,147 @@ ignore = [  # NOTE: non-permeanent exclusions should be added to `.ruff.toml` in
 [tool.ruff.pydocstyle]
 convention = "numpy"
 
+[tool.pytest.ini_options]
```
Comment (mhvk) [inline pyproject.toml:268]: This I don't understand; either be alphabetical or not - why is this moved past `tools.ruff`? Certainly, `pytest` is far more important to see than `ruff`.

**Review by mhvk — COMMENTED** [review]
Somewhat mixed about it changing ordering in what seems an arbitrary fashion. I'd like the first astropy keyword to be "astronomy", not "ascii".

p.s. I find `pre-commit` a bit of a pain in slowing down development; this presumably is very fast if `pyproject.toml` is not touched?

File: pyproject.toml [hunk 5]
```diff
@@ -1,110 +1,110 @@
 [project]
 name = "astropy"
-dynamic = [
-    "version"
-]
 description = "Astronomy and astrophysics core library"
 readme = "README.rst"
+keywords = [
```
Comment (mhvk) [inline pyproject.toml:5]: The alphabetization has removed meaning here. `astropy` is not first and foremost about `ascii` (not sure why that was put on there, though I guess table reading...).

**Comment by nstarman** [comment]
> p.s. I find pre-commit a bit of a pain in slowing down development; this presumably is very fast if pyproject.toml is not touched?

Interesting. What step in pre-commit is slow?

**Comment by WilliamJamieson** [comment]
The remove star is slow.

**Comment by nstarman** [comment]
On `run` or only `run -a`?

**Comment by nstarman** [comment]
I find it a bit slow on `run -a`, but not `run` (unless I've edited every file in GH, of course)

**Comment by nstarman** [comment]
Ping @Saransh-cpp for this discussion of speed.

**Comment by mhvk** [comment]
It is the collection. And, no, it is not that slow, it is just enough to be irritating, that's why I'm a bit wary of adding more steps. (Plus I fairly regularly clean up the repository with `git clean -fxd`, and then  one has to reinstall, but that's arguably on me.)

**Comment by nstarman** [comment]
> It is the collection. And, no, it is not that slow, it is just enough to be irritating

Interesting. I'm on an M2 MBPro and it's fast enough I don't notice.
I wonder what steps are slow, maybe black & removestar? They have to build an AST in Python. The rest are regex or rust

**Comment by WilliamJamieson** [comment]
removestar is noticeably slow for me.

**Comment by nstarman** [comment]
We could move removestar to a cron job, or if https://github.com/asmeurer/removestar/issues/51 (or similar) can speed it up...

**Comment by Saransh-cpp** [comment]
Thanks for the ping!

Re pyproject-fmt:

The hook was not added to scientific-python's development guide because it was not very flexible - https://github.com/scikit-hep/scikit-hep.github.io/pull/214 ([scikit-hep.github.io](https://scikit-hep.org) was ported to create [learn.scientific-python.org/development](https://learn.scientific-python.org/development/)). I've also discovered some bugs in the tool in the past, and though I planned to work on them, I never got the time :(

(not completely against the hook, some of my repositories use it, some of them don't, and I really appreciate the work done on it, everything depends on how well it fits within astropy!)

Re removestar's speed:

I can shift it to the cron job until I profile and identify the bottleneck. This might take some time, and I am not sure when I'll start working on it 🙂
